lib/charms/operator_libs_linux/v0/dnf.py

```diff
@@ -0,0 +1,462 @@
+        return self.__data["name"]
+
+    @property
+    def arch(self) -> Optional[str]:
+        """Get architecture of package.
+
+        Returns:
+            (Optional[str]):
+                Architecture of package. Defaults to None if package is absent.
+        """
+        return self.__data.get("arch", None)
+
+    @property
+    def epoch(self) -> Optional[str]:
+        """Get epoch of package.
+
+        Returns:
+            (Optional[str]):
+                Epoch of package. Defaults to None if package is absent.
+        """
+        return self.__data.get("epoch", None)
+
+    @property
+    def version(self) -> Optional[str]:
+        """Get version of package.
+
+        Returns:
+            (Optional[str]):
+                Version number of package. Defaults to None if package is absent.
+        """
+        return self.__data.get("version", None)
+
+    @property
+    def fullversion(self) -> Optional[str]:
+        """Get full version of package.
+
+        Returns:
+            (Optional[str]):
+                Full version of package. Defaults to None if package is absent.
+        """
+        if self.absent:
+            return None
+
+        return (
+            f"{(f'{self.epoch}:' if self.epoch else '')}"
+            f"{self.version}{(f'-{self.release}' if self.release else '')}"
+        )
+
+    @property
+    def release(self) -> Optional[str]:
+        """Get release of package.
+
+        Returns:
+            (Optional[str]):
+                Release of package. Defaults to None if package is absent.
+        """
+        return self.__data.get("release", None)
+
+    @property
+    def repo(self) -> Optional[str]:
+        """Get repository package is from.
+
+        Returns:
+            (Optional[str]):
+                Repository of package: Defaults to None if package is absent.
+        """
+        return self.__data.get("repo", None)
+
+
+def _load_cache(
+    cache_type: Literal["installed", "available"]
+) -> List[Dict[str, Union[str, _PackageState]]]:
+    """Load DNF cache information.
+
+    Args:
+        cache_type (Literal["installed", "available"]):
+            Cache information category to load.
+
+    Raises:
+        DNFExecutionError: Raised if incorrect cache type is passed.
+
+    Returns:
+        (List[Dict[str, Union[str, _PackageState]]]):
+            Loaded cache information.
+    """
+    if cache_type not in ["installed", "available"]:
+        raise DNFExecutionError(
+            f"Cache type must be either installed or available, not {cache_type}"
+        )
+
+    result = []
+    state = {"installed": _PackageState.INSTALLED, "available": _PackageState.AVAILABLE}
+    for pkg_name, pkg_version, pkg_repo in itertools.filterfalse(
+        lambda x: len(x) != 3,
+        [
+            line.strip("\n").strip().split()
+            for line in StringIO(
+                _dnf("list", postargs=[f"--{cache_type}"]).stdout.strip("\n"),
+            )
+        ],
+    ):
```

Comment:
How does a comment block above this loop sound that walks through the steps? It might still be confusing even if I break it up slightly :sweat_smile:

Essentially I am:

1. Grabbing the dnf list of packages marked a certain may (installed or available).
2. Stripping the extra newline characters surrounding the captured output.
3. Wrapping the captured output with StringIO so I can iterate over the output like a file.
4. Strip newline characters, strip extra whitespace, and then split by whitespace.
5. Use `itertools.filterfalse(..., ...)` to remove splits that are not `len(...) == 3`. This gets rid of the "dnf metadata check" and "here are the types of packages I just" listed methods.

Reply:
I had the exact same comment as Jon below. I think part of the issue here is how the Black formatter is breaking things up (sometimes it's annoying like that), but part of it is the nested-ness of it, and the fact that the "main call", the `_dnf()` call, is kind of hidden away nested in the middle of things.

Comment blocks are useful when necessary, but I think here it'd be unnecessary if we wrote the simpler, non-nested version of the code (one of the versions I've shown below).

Reply:
Dropped in https://github.com/canonical/operator-libs-linux/pull/75/commits/661114b536c049933697507536f6270a1d2a6beb